core/templates/pages/exploration-editor-page/modal-templates/state-diff-modal.controller.spec.ts

```diff
@@ -0,0 +1,138 @@
+        .and.returnValue($q.resolve('sample-csrf-token'));
+
+      newState = StateObjectFactory.createDefaultState(newState);
+      oldState = StateObjectFactory.createDefaultState(oldState);
+
+      $uibModalInstance = jasmine.createSpyObj(
+        '$uibModalInstance', ['close', 'dismiss']);
+
+      // There are two post requests with the same endpoint.
+      $httpBackend.expectPOST('/createhandler/state_yaml/exp1').respond({
+        yaml: 'Yaml data'
+      });
+      $httpBackend.expectPOST('/createhandler/state_yaml/exp1').respond({
+        yaml: 'Yaml data'
+      });
+
+      $scope = $rootScope.$new();
+      $controller('StateDiffModalController', {
+        $scope: $scope,
+        $uibModalInstance: $uibModalInstance,
+        headers: headers,
+        newState: newState,
+        newStateName: newStateName,
+        oldState: oldState,
+        oldStateName: oldStateName
+      });
+    }));
+
+    it('should evaluate scope variables values correctly', function() {
+      expect($scope.headers).toBe(headers);
+      expect($scope.newStateName).toBe(newStateName);
+      expect($scope.oldStateName).toBe(oldStateName);
+      expect($scope.yamlStrs).toEqual({});
+    });
+
+    it('should evaluate yaml strings object', function() {
+      $httpBackend.flush();
+      expect($scope.yamlStrs.leftPane).toBe('Yaml data');
+      expect($scope.yamlStrs.rightPane).toBe('Yaml data');
+    });
+  });
+
+  describe('when new state and old state are falsy', function() {
+    newState = null;
+    oldState = null;
+
+    beforeEach(angular.mock.inject(function($injector, $controller) {
+      $q = $injector.get('$q');
+      var $rootScope = $injector.get('$rootScope');
+      $timeout = $injector.get('$timeout');
+      ContextService = $injector.get('ContextService');
+      spyOn(ContextService, 'getExplorationId').and.returnValue(explorationId);
+      CsrfTokenService = $injector.get('CsrfTokenService');
+
+      spyOn(CsrfTokenService, 'getTokenAsync')
+        .and.returnValue($q.resolve('sample-csrf-token'));
+
+      $uibModalInstance = jasmine.createSpyObj(
+        '$uibModalInstance', ['close', 'dismiss']);
+
+      $scope = $rootScope.$new();
+      $controller('StateDiffModalController', {
+        $scope: $scope,
+        $uibModalInstance: $uibModalInstance,
+        headers: headers,
+        newState: newState,
+        newStateName: newStateName,
+        oldState: oldState,
+        oldStateName: oldStateName
+      });
+    }));
+
+    it('should evaluate scope variables values correctly', function() {
+      expect($scope.headers).toBe(headers);
+      expect($scope.newStateName).toBe(newStateName);
+      expect($scope.oldStateName).toBe(oldStateName);
+      expect($scope.yamlStrs).toEqual({});
+    });
+
+    it('should evaluate yaml strings object', function() {
+      // Flush 400 milliseconds $timeout in order to set right and left panes.
+      $timeout.flush(400);
```

Comment:
@marianazangrossi @nithusha21 This test is flaky. It is failing on #9714, which is unrelated. The stack trace is:

```
	Error: Expected undefined to be ''.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:51850:46)
	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:285902:30)
	    at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:482907:43)
	Error: Expected undefined to be ''.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:51851:47)
	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:285902:30)
	    at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:482907:43)
Chrome Headless 83.0.4103.116 (Linux x86_64): Executed 495 of 2885 (1 FAILED) (0 secs / 7.072 secs)
Chrome Headless 83.0.4103.116 (Linux x86_64) State Diff Modal Controller when new state and old state are falsy should evaluate yaml strings object FAILED
	Error: Expected undefined to be ''.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:51850:46)
	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:285902:30)
	    at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:482907:43)
	Error: Expected undefined to be ''.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:51851:47)
	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:285902:30)
	    at ProxyZoneSpec../node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:482907:43)
```

Given this, I am reverting this PR.

Please take the following next steps:

(a) Remove $timeout.flush from the frontend tests. It leads to flakiness. Please fix this elsewhere too.
(b) Update the wiki page and add a lint check to ban the use of this construct in frontend tests. Provide an error message in the lint check to explain why this is not a recommended practice.

Reply:
Hi @seanlip, I think removing `$timeout.flush` is not the ideal solution. When removing it, there is no way to reach out to the code inside `$timeout` callback. There are another approaches to deal with timing behavior but they are confusing and more likely to fail, like [jasmine clock](https://jasmine.github.io/api/2.7/Clock.html). Another approach is to change the `$timeout `to `setTimeout` which will lead to the same situation: we will need to create a `setTimeout` inside de spec with specific milliseconds and with `done` parameter in order to test the callback. Which may cause the same error due to milliseconds waited.

What I'd suggest in this case is when testing timing behavior, always wait twice as long. For example: if the callback is going to be triggered after 200 msec, we are going to flush the `$timeout` for 400 msec. This way we're going to be sure that the callback is going to be executed. I've faced this error before and I think it happens because sometimes the callback is triggered some milliseconds after the flush. Waiting for more time will fix the problem without needing to change all the specs that use it.

If both of you agree with this, I can include this solution on the wiki page.

Reply:
Just an update: it looks like this specific file is very tricky to test. Trying to work with double the time is not the solution here. However, I don't think the problem is the `$timeout.flush`, I've been trying to test with other approaches and it's not working either. I don't know why it's failing but I'll go deeper to find a solution.

Reply:
Perhaps in the main codebase you can do something like "waitfor codemirror to load" instead?